kroxylicious-kms-provider-aws-kms/src/main/java/io/kroxylicious/kms/provider/aws/kms/credentials/Ec2MetadataCredentialsProvider.java

```diff
@@ -0,0 +1,301 @@
+        if (witness == null) {
+            // there's no current credential, let's create one
+            executorService.execute(() -> refreshCredential(newCredFuture));
+            return newCredFuture.minimalCompletionStage();
+        }
+        else if (isExpired(witness) || witness.isCompletedExceptionally()) {
+            // current credential is expired, or it has been completed exceptionally.
+            // throw it away and generate a new one.
+            // we don't normally expect to follow the expired path as the preemptive refresh ought to have
+            // caused its refresh before its expiration.
+            current.compareAndSet(witness, null);
+            return getCredentials();
+        }
+
+        return witness.minimalCompletionStage();
+    }
+
+    private void scheduleCredentialRefresh(Instant credentialExpiration) {
+        var delay = (long) Math.max(0, this.lifetimeFactor * (credentialExpiration.toEpochMilli() - systemClock.millis()));
+        LOGGER.debug("Scheduling preemptive refresh of AWS credentials in {}ms", delay);
+
+        var refreshedCredFuture = new CompletableFuture<SecurityCredentials>();
+        executorService.schedule(() -> {
+            refreshCredential(refreshedCredFuture);
+            refreshedCredFuture.thenApply(sc -> {
+                var previous = current.getAndSet(refreshedCredFuture);
+                // the previous future have been already complete, but for safety complete it anyway.
+                previous.complete(sc);
+                return null;
+            });
+        }, delay, TimeUnit.MILLISECONDS);
+    }
+
+    private boolean isExpired(CompletableFuture<SecurityCredentials> witness) {
+        if (witness.isDone() && !witness.isCompletedExceptionally()) {
+            try {
+                return Optional.ofNullable(witness.getNow(null))
+                        .map(SecurityCredentials::expiration)
+                        .map(exp -> systemClock.instant().isAfter(exp))
+                        .orElse(false);
+            }
+            catch (CancellationException | CompletionException e) {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    private void refreshCredential(CompletableFuture<SecurityCredentials> future) {
+        getToken()
+                .thenCompose(tokenResponse -> client.sendAsync(createSecurityCredentialsRequest(tokenResponse.body()), HttpResponse.BodyHandlers.ofByteArray()))
+                .thenApply(Ec2MetadataCredentialsProvider::checkResponseStatus)
+                .thenApply(HttpResponse::body)
+                .thenApply(this::toSecurityCredentials)
+                .thenApply(this::checkSuccessfulState)
+                .whenComplete((credentials, t) -> propagateResultToFuture(credentials, t, future));
+    }
+
+    private void propagateResultToFuture(SecurityCredentials credentials, Throwable t, CompletableFuture<SecurityCredentials> target) {
```

Comment:
I've reorder so that it is clear that the target future will always be completed with a value or exceptionally.  There was also a muddle with the `lifetimeFactor` which I sorted out.